src/main/java/com/jcraft/jsch/Session.java

```diff
@@ -2101,7 +2160,26 @@ public void disconnect() {
      * } }
      */
 
-    Channel.disconnect(this);
+    List<Channel> channelsCopy;
+    Lock l = channelsLock.writeLock();
+    l.lock();
+    try {
+      disconnectingChannels = true;
+      channelsCopy = new ArrayList<>(channels);
+    } finally {
+      l.unlock();
+    }
+
+    for (Channel c : channelsCopy) {
+      try {
+        c.disconnect();
+      } catch (Exception e) {
+        if (getLogger().isEnabled(Logger.DEBUG)) {
+          getLogger().log(Logger.DEBUG, "Exception disconnecting channel of type "
+              + Util.byte2str(c.type) + ": " + e.getMessage(), e);
+        }
```

Comment:
I've tweaked this to catch any exceptions while disconnecting individual channels so that it doesn't prematurely halt disconnection of the session itself.